### PR TITLE
Parameterize HandleAllocator and use it in VulkanDriver.

### PR DESCRIPTION
**Pull request by prideout**

### CODE REVIEW

**Comment by prideout** [comment]
Please review CL's individually.

File: filament/backend/include/private/backend/HandleAllocator.h [hunk 1]
```diff
@@ -36,6 +36,7 @@ namespace filament::backend {
 /*
  * A utility class to efficiently allocate and manage Handle<>
  */
+template <int P0, int P1, int P2>
```
Comment (pixelflinger) [inline filament/backend/include/private/backend/HandleAllocator.h:39]: use `size_t` instead of `int` (or at least `unsigned`)

File: filament/backend/src/HandleAllocator.cpp [hunk 2]
```diff
@@ -24,26 +24,29 @@ namespace filament::backend {
...
+template <int P0, int P1, int P2>
 UTILS_NOINLINE
-HandleAllocator::Allocator::Allocator(AreaPolicy::HeapArea const& area)
+HandleAllocator<P0, P1, P2>::Allocator::Allocator(AreaPolicy::HeapArea const& area)
         : mArea(area) {
     // TODO: we probably need a better way to set the size of these pools
     const size_t unit = area.size() / 32;
     const size_t offsetPool1 =      unit;
     const size_t offsetPool2 = 16 * unit;
     char* const p = (char*)area.begin();
```
Comment (pixelflinger) [inline filament/backend/src/HandleAllocator.cpp:35]: We probably also need some parameters to define the relative size of each pool arena, can be in another PR.

File: filament/backend/src/opengl/OpenGLDriver.h [hunk 3]
```diff
@@ -249,7 +249,8 @@ class OpenGLDriver final : public backend::DriverBase {
 
     // Memory management...
 
-    backend::HandleAllocator mHandleAllocator;
+    // See also the explicit template instantiation in HandleAllocator.cpp
+    backend::HandleAllocator<16, 64, 208> mHandleAllocator;
```
Comment (pixelflinger) [inline filament/backend/src/opengl/OpenGLDriver.h:253]: You could define an alias inside the `#if FILAMENT_SUPPORTS_FOO`, for e.g.
`using HandleAllocatorGL = HandleAllocator<16,64,208>;`
So that we don't have to repeat the sizes here?

**Review by pixelflinger — CHANGES_REQUESTED** [review]
very nice

File: filament/backend/src/opengl/OpenGLDriver.h (hunk 3 above)
Comment (prideout) [reply filament/backend/src/opengl/OpenGLDriver.h:253]: Yeah but the `#if FILAMENT_SUPPORTS_FOO` thingie is in a cpp file, and C++ doesn't let you instantiate a template using an alias, as far as I can tell.

Comment (prideout) [reply filament/backend/src/opengl/OpenGLDriver.h:253]: Maybe I could do this with a macro...

File: filament/backend/include/private/backend/HandleAllocator.h (hunk 1 above)
Comment (prideout) [reply filament/backend/include/private/backend/HandleAllocator.h]: Done.

File: filament/backend/src/opengl/OpenGLDriver.h (hunk 3 above)
Comment (prideout) [reply filament/backend/src/opengl/OpenGLDriver.h]: Done.

**Review by pixelflinger — APPROVED** [review]
